packages/platform-android/native_modules.gradle

```diff
@@ -0,0 +1,250 @@
+  }
+
+  /**
+   * Include the react native modules android projects and specify their project directory
+   */
+  void addReactNativeModuleProjects() {
+    reactNativeModules.forEach { reactNativeModule ->
+      String name = reactNativeModule["name"]
+      String androidSourceDir = reactNativeModule["androidSourceDir"]
+      defaultSettings.include(":${name}")
+      defaultSettings.project(":${name}").projectDir = new File("${androidSourceDir}")
+    }
+  }
+
+  /**
+   * Adds the react native modules as dependencies to the users `app` project
+   */
+  void addReactNativeModuleDependencies() {
+    reactNativeModules.forEach { reactNativeModule ->
+      def name = reactNativeModule["name"]
+      project.dependencies {
+        // TODO(salakar): are other dependency scope methods such as `api` required?
+        implementation project(path: ":${name}")
+      }
+    }
+  }
+
+  /**
+   * This returns the users project root (e.g. where the node_modules dir is located).
+   *
+   * This defaults to up one directory from the root android directory unless the user has defined
+   * a `ext.reactNativeProjectRoot` extension property
+   *
+   * @return
+   */
+  File getReactNativeProjectRoot() {
+    if (this.extension.has("reactNativeProjectRoot")) {
+      File rnRoot = File(this.extension.get("reactNativeProjectRoot"))
+      // allow custom React Native project roots for non-standard directory structures
+      this.logger.debug("${LOG_PREFIX}Using custom React Native project root path '${rnRoot.toString()}'")
+      return rnRoot
+    }
+
+    File androidRoot
+
+    if (this.project) {
+      androidRoot = this.project.rootProject.projectDir
+    } else {
+      androidRoot = this.defaultSettings.rootProject.projectDir
+    }
+
+    this.logger.debug("${LOG_PREFIX}Using default React Native project root path '${androidRoot.parentFile.toString()}'")
+    return androidRoot.parentFile
+  }
+
+  /**
+   * Code-gen a java file with all the detected ReactNativePackage instances automatically added
+   *
+   * @param outputDir
+   * @param generatedFileName
+   * @param generatedFileContentsTemplate
+   * @param applicationId
+   */
+  void generatePackagesFile(File outputDir, String generatedFileName, String generatedFileContentsTemplate, String applicationId) {
+    ArrayList<HashMap<String, String>>[] packages = this.reactNativeModules
+
+    String packageImports = ""
+    String packageClassInstances = ""
+
+    if (packages.size() > 0) {
+      packageImports = "import ${applicationId}.BuildConfig;\n\n"
+      packageImports = packageImports + packages.collect {
+        "// ${it.name}\n${it.packageImportPath}"
+      }.join(';\n')
+      packageClassInstances = ",\n      " + packages.collect { it.packageInstance }.join(',')
+    }
+
+    String generatedFileContents = generatedFileContentsTemplate
+      .replace("{{ packageImports }}", packageImports)
+      .replace("{{ packageClassInstances }}", packageClassInstances)
```

Comment:
I am wondering if there's any way we can wrap an instance of `ReactPackage` here to `LazyReactPackage`. By default, `packageClassInstance` is e.g. `new CodePushPackage()` and in order to convert it to a LazyPackage, we would have to update that package itself.

Is there a way to wrap it or create our custom lazy package for each ReactPackage, that will lazily create it and return everything what's needed?

Reply:
I think `CodePushPackage` class itself would need to extend `LazyReactPackage` instead of `ReactPackage` for this to work, not something we can do? 🤷‍♂️ 

Forcing every package to be lazy will also cause issues for packages that are not inherently meant to be lazy, e.g. modules that have app initialization logic that must always be run?